File: 780_Remove_Invalid_Parentheses.py
#!/usr/bin/python3

'''
780. Remove Invalid Parentheses
Remove the minimum number of invalid parentheses in order to make the input string valid. Return all possible results.

Example
"()())()" -> ["()()()", "(())()"]
"(a)())()" -> ["(a)()()", "(a())()"]
")(" -> [""]


Author: Yuan Li
Date: 6/17/2018 
Difficulty: Hard
'''

class Solution:
    """
    @param s: The input string
    @return: Return all possible results
    """
    def removeInvalidParentheses(self, s):
        # Write your code here
        res = []
        res = self.helper(res, s, len(s))
        
        if res == []:
            return [""]
        return res
        
    def helper(self, res, s, n):
        if n == 0:
            return res
        
        if self.isValidParentheses(s):
            if s not in res:
                res.append(s)
            return res
         
        l, r, pr = 0, 0, []
        for i in range(n):
            l += s[i] == '('
            if s[i] == ')':
                r += 1
                pr.append(i)
            
            if l < r:
                for j in pr:
                    subs = s[:j] + s[j+1:] 
                    res = self.helper(res, subs, n - 1)
                break

        l, r, pl = 0, 0, []
        for i in range(n):
            if s[n - 1 - i] == '(':
                l += 1
                pl.append(n - 1 - i)
            r += s[n - 1 - i] == ')'
            
            if l > r:
                for j in pl:
                    subs = s[:j] + s[j+1:]
                    res = self.helper(res, subs, n - 1)
                break
        
        return res
        
    def isValidParentheses(self, s):
        if not s:
            return True
        
        stack = []
        
        for c in s:
            if not stack: 
                if c == '(' or c == ')':
                    stack.append(c)
                continue
            
            if c == ')' and stack[-1] == '(':
                stack.pop()
                
            elif c == '(' or c == ')':
                stack.append(c)
                
        if stack:
            return False
            
        return True

File: 780_Remove_Invalid_Parentheses.cpp
/*
780. Remove Invalid Parentheses
Remove the minimum number of invalid parentheses in order to make the input string valid. Return all possible results.

The input string may contain letters other than the parentheses ( and ).

Example
"()())()" -> ["()()()", "(())()"]
"(a)())()" -> ["(a)()()", "(a())()"]
")(" -> [""]


Author: Yuan Li
Date: 7/16/2018 
Difficulty: Hard
*/

class Solution {
public:
    /**
     * @param s: The input string
     * @return: Return all possible results
     */
    vector<string> removeInvalidParentheses(string &s) {
        // Write your code here
        unordered_set<string> combs;
        string comb(s);
        
        int maxK = count(s.begin(), s.end(), '(') + count(s.begin(), s.end(), ')');
        for (int k = 0; k <= maxK; ++k) {
            dfs(k, comb, 0, combs);
            
            if (!combs.empty())
                break;
        }
        
        vector<string> res(combs.begin(), combs.end()); 
        return res;
    }
    
    void dfs(int remainK, string comb, int idx, 
            unordered_set<string> &combs) {
        if (remainK == 0) {
            //if (isValidParentheses(comb)) {
            if (isValidParentheses(comb, comb.length())) {
                combs.insert(comb);
            }
            
            return;
        }
        
        //if (idx == comb.length() || comb.length() - idx < remainK)
        if (idx == comb.length() || comb.length() - idx < remainK || 
            !isValidParentheses(comb, idx))
            return;
            
        char p = comb[idx];
        if (p == '(' || p == ')') {
            // remove this parenthese
            comb.erase(idx, 1);
            dfs(remainK - 1, comb, idx, combs);
            
            // do not remove this parenthese 
            comb.insert(idx, 1, p);
            dfs(remainK, comb, idx + 1, combs);
        } else { // p is a letter
            dfs(remainK, comb, idx + 1, combs);
        }
    }
    
    bool isValidParentheses(string &s) {
        int cntL = 0, cntR = 0;
        for (char c : s) {
            if (c == '(') {
                cntL++;
            } else if (c == ')') {
                cntR++;
            }
            
            if (cntR > cntL)
                return false;
        }
        
        cntL = 0; cntR = 0;
        for (int i = s.length() - 1; i >= 0; --i) {
            char c = s[i];
            if (c == '(') {
                cntL++;
            } else if (c == ')') {
                cntR++;
            }
            
            if (cntL > cntR)
                return false;
        }
        
        return true;
    }
    
    // optimized isValidParentheses
    bool isValidParentheses(string &s, int idx) {
        int cntL = 0, cntR = 0;
        for (int i = 0; i < idx; ++i) {
            char c = s[i];
            if (c == '(') {
                cntL++;
            } else if (c == ')') {
                cntR++;
            }
            
            if (cntR > cntL)
                return false;
        }
    
        if (idx < s.length())
            return true;
            
        cntL = 0; cntR = 0;
        for (int i = s.length() - 1; i >= 0; --i) {
            char c = s[i];
            if (c == '(') {
                cntL++;
            } else if (c == ')') {
                cntR++;
            }
            
            if (cntL > cntR)
                return false;
        }
        
        return true;
    }
};
